File: src/main/java/me/ste/stevesseries/base/ChatInputListener.java
package me.ste.stevesseries.base;

import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.Listener;
import org.bukkit.event.player.AsyncPlayerChatEvent;
import org.bukkit.event.player.PlayerCommandPreprocessEvent;

import java.util.concurrent.CompletableFuture;

public class ChatInputListener implements Listener {
    @EventHandler(priority = EventPriority.HIGHEST)
    public void onPlayerCommandPreprocess(PlayerCommandPreprocessEvent event) {
        if(ChatInputManager.getPlayerMap().containsKey(event.getPlayer().getUniqueId())) {
            for(CompletableFuture<String> future : ChatInputManager.getPlayerMap().get(event.getPlayer().getUniqueId())) {
                future.complete(event.getMessage());
            }
            ChatInputManager.getPlayerMap().removeAll(event.getPlayer().getUniqueId());
            event.setCancelled(true);
        }
    }

    @EventHandler(priority = EventPriority.HIGHEST)
    public void onPlayerChat(AsyncPlayerChatEvent event) {
        if(ChatInputManager.getPlayerMap().containsKey(event.getPlayer().getUniqueId())) {
            for(CompletableFuture<String> future : ChatInputManager.getPlayerMap().get(event.getPlayer().getUniqueId())) {
                future.complete(event.getMessage());
            }
            ChatInputManager.getPlayerMap().removeAll(event.getPlayer().getUniqueId());
            event.setCancelled(true);
        }
    }
}

File: src/main/kotlin/me/ste/stevesseries/base/ChatInputListener.kt
package me.ste.stevesseries.base

import me.ste.stevesseries.base.BaseAPIImpl
import me.ste.stevesseries.base.api.event.EventManager
import org.bukkit.event.Listener
import org.bukkit.event.player.AsyncPlayerChatEvent
import org.bukkit.event.player.PlayerCommandPreprocessEvent

class ChatInputListener(
    private val impl: BaseAPIImpl
) : Listener {
    fun onPlayerChat(event: AsyncPlayerChatEvent) {
        if (this.impl.consumeChatInput(event.player, event.message)) {
            event.isCancelled = true
        }
    }

    fun onPlayerCommand(event: PlayerCommandPreprocessEvent) {
        if (this.impl.consumeChatInput(event.player, event.message)) {
            event.isCancelled = true
        }
    }

    fun register(manager: EventManager) {
        manager.listen(listener = this::onPlayerChat)
        manager.listen(listener = this::onPlayerCommand)
    }
}